### PR TITLE
Introduce metrics module / implement cognitive complexity

### PR DESCRIPTION
**Pull request by arturbosch**
Unlike Cyclomatic complexity, which is a measurement for the testability of a code unit, cognitive complexity is a measurement for the maintainability of a code unit.
It tells the user how difficult a piece of code is to read and understand.

Reference used: https://www.sonarsource.com/docs/CognitiveComplexity.pdf

This is the first step towards a metric module.
Other steps are:

- Moving the processors of `detekt-core` which compute a `ProjectMetric` into detekt-metrics
- Including cognitive complexity into our complexity report
- (To discuss) Allowing to measure cognitive instead of cyclomatic complexity for the `ComplexMethod` rule (A non breaking way for 1.X.X would be to implement the switch via a config property `algorithm: cognitive`)

### CODE REVIEW

**Comment by schalkms** [comment]
I'll review this tomorrow.

File: detekt-metrics/src/main/kotlin/io/github/detekt/metrics/CognitiveComplexity.kt [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        private fun calculateBinaryExprComplexity(usedExpr: List<BinExprHolder>) {
+            var lastOp: IElementType? = null
+            for (binExpr in usedExpr) {
+                if (lastOp == null) {
+                    complexity++
+                } else {
+                    if (lastOp != binExpr.op || binExpr.isEnclosed) {
+                        complexity++
+                    }
+                }
```
Comment (schalkms) [inline detekt-metrics/src/main/kotlin/io/github/detekt/metrics/CognitiveComplexity.kt:172]:
```suggestion
                } else if (lastOp != binExpr.op || binExpr.isEnclosed) {
                    complexity++
                }
```

File: detekt-metrics/src/test/kotlin/io/github/detekt/metrics/CognitiveComplexitySpec.kt [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+            assertThat(CognitiveComplexity.calculate(code)).isEqualTo(2)
+        }
+
+        it("adds nesting for nested functions but not complexity") {
+            val code = compileContentForTest("""
+                fun main() { fun run() { if (true) {} } }
+            """)
+
+            assertThat(CognitiveComplexity.calculate(code)).isEqualTo(2)
+        }
```
Comment (schalkms) [inline detekt-metrics/src/test/kotlin/io/github/detekt/metrics/CognitiveComplexitySpec.kt:136]: It's a bit hard to understand why these two have the mentioned complexity value.

Comment (schalkms) [reply detekt-metrics/src/test/kotlin/io/github/detekt/metrics/CognitiveComplexitySpec.kt:136]: Maybe you can enlighten me. ;-)

File: detekt-metrics/src/test/kotlin/io/github/detekt/metrics/CognitiveComplexitySpec.kt [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+                                !cond 
+                                && !(cond && cond && cond)  // +2
+                            ) {}
+                        }
+                    """)
+
+                    assertThat(CognitiveComplexity.calculate(code)).isEqualTo(3)
+                }
+
+                it("adds one for negated like operator but only once") {
```
Comment (schalkms) [inline detekt-metrics/src/test/kotlin/io/github/detekt/metrics/CognitiveComplexitySpec.kt:215]: This has the same test description as above. If I recall it correctly, the Spek framework had problem with that in the past.

**Comment by schalkms** [comment]
> (To discuss) Allowing to measure cognitive instead of cyclomatic complexity for the ComplexMethod rule (A non breaking way for 1.X.X would be to implement the switch via a config property algorithm: cognitive)

I'd leave this for now as it is and wait for feedback. This is maybe a topic for detekt > v2.

File: detekt-metrics/src/test/kotlin/io/github/detekt/metrics/CognitiveComplexitySpec.kt (hunk 2 above)
Comment (arturbosch) [reply detekt-metrics/src/test/kotlin/io/github/detekt/metrics/CognitiveComplexitySpec.kt:136]: Cognitive complexity adds +1 for each nesting level. Lambdas and nested function increase the nesting level but do not add complexity.
In this cases we have an `if` which would normally add +1 but due to nesting level 1, it adds +2.
hope that helps

File: detekt-metrics/src/test/kotlin/io/github/detekt/metrics/CognitiveComplexitySpec.kt (hunk 3 above)
Comment (arturbosch) [reply detekt-metrics/src/test/kotlin/io/github/detekt/metrics/CognitiveComplexitySpec.kt:215]: Thx! fixed.

**Pull request merged**